### PR TITLE
tools: Remove contrib from the VSCode code completion db

### PR DESCRIPTION
**Pull request by abeyad**
The contrib directory sometimes doesn't build and prevents the VSCode code completion script
(tools/vscode/refresh_compdb.sh) from generating the symbols database.  This change overrides
the default directories built for the VSCode script by removing contrib and only building source,
test, and tools by default.

To allow including the contrib directories in the VSCode completion database, remove the 
`--exclude_contrib` flag from refresh_compdb.sh.

Signed-off-by: Ali Beyad <abeyad@google.com>

### CODE REVIEW

File: tools/vscode/refresh_compdb.sh [hunk 1]
```diff
@@ -6,7 +6,7 @@ bazel_or_isk=bazelisk
 command -v bazelisk &> /dev/null || bazel_or_isk=bazel
 
 # Setting TEST_TMPDIR here so the compdb headers won't be overwritten by another bazel run
-TEST_TMPDIR=${BUILD_DIR:-/tmp}/envoy-compdb tools/gen_compilation_database.py --vscode --bazel=$bazel_or_isk
+TEST_TMPDIR=${BUILD_DIR:-/tmp}/envoy-compdb tools/gen_compilation_database.py --vscode --bazel=$bazel_or_isk //source/... //test/... //tools/...
```
[the remark this answers is not in the document]
Comment (abeyad) [reply tools/vscode/refresh_compdb.sh:9]: Good point, they weren't getting compiled before in the VSCode symbols db but I added it and it works, so thanks for the suggestion!  Just uploaded a new commit that adds both.

**Review by RyanTheOptimist — APPROVED** [review]

File: tools/vscode/refresh_compdb.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (lizan) [reply tools/vscode/refresh_compdb.sh]: Here is the default in the script, you don't need //envoy or //api becasue they are pulled by dependencies anyway. https://github.com/envoyproxy/envoy/blob/main/tools/gen_compilation_database.py#L122

**Comment by daixiang0** [comment]
It confused me, since CI would build contrib, if failed to build, it can not land I think.

**Comment by abeyad** [comment]
> It confused me, since CI would build contrib, if failed to build, it can not land I think.

I found it confusing too, but running refresh_compdb.sh has been failing for a few weeks for me in the same place:
```
ERROR: /usr/local/google/home/abeyad/dev/envoy/contrib/hyperscan/matching/input_matchers/source/BUILD:21:12: Foreign Cc - CMake: Building hyperscan failed: (Exit 1): bash failed: error executing command (from target //contrib/hyperscan/matching/input_matchers/source:hyperscan) /bin/bash -c bazel-out/k8-fastbuild/bin/contrib/hyperscan/matching/input_matchers/source/hyperscan_foreign_cc/wrapper_build_script.sh

Use --sandbox_debug to see verbose messages from the sandbox
rules_foreign_cc: Build failed!
rules_foreign_cc: Keeping temp build directory and dependencies directory for debug.
rules_foreign_cc: Please note that the directories inside a sandbox are still cleaned unless you specify --sandbox_debug Bazel command line flag.
rules_foreign_cc: Printing build logs:
```

I don't know anything about the hyperscan lib and am new to the Envoy project.  Is this something you've seen before?

**Comment by daixiang0** [comment]
Actually I do not met that, could you share detailed build log here?

/cc @zhxie

**Comment by abeyad** [comment]
I pasted the logs to this file: https://github.com/abeyad/misc/blob/main/envoy_hyperscan_build.log

**Comment by zhxie** [comment]
> I don't know anything about the hyperscan lib and am new to the Envoy project. Is this something you've seen before?

It is just caused by your compiler or compiler options. It will work as expected if you are in `./ci/run_envoy_docker.sh` environment.

**Comment by zhxie** [comment]
BTW, I do not think remove contrib directly is a good idea. But I agree with that we can remove contrib by default, and add some options like `tools/vscode/refresh_compdb.sh --contrib` to build compilation db for both source and contrib.

**Comment by abeyad** [comment]
> BTW, I do not think remove contrib directly is a good idea. But I agree with that we can remove contrib by default, and add some options like `tools/vscode/refresh_compdb.sh --contrib` to build compilation db for both source and contrib.

Thanks for your feedback!  This is a good idea.  I pushed a commit that adds a `--include-contrib` option.  Please take another look.

**Comment by lizan** [comment]
> I pasted the logs to this file: https://github.com/abeyad/misc/blob/main/envoy_hyperscan_build.log

Seems related to your environment

```
/usr/lib/llvm-13/bin/clang  -I/tmp/envoy-compdb/_bazel_abeyad/b3914d36ee2d69a6c5998cfece95c54a/sandbox/linux-sandbox/3/execroot/envoy/bazel-out/k8-fastbuild/bin/contrib/hyperscan/matching/input_matchers/source/hyperscan.build_tmpdir/util -I/tmp/envoy-compdb/_bazel_abeyad/b3914d36ee2d69a6c5998cfece95c54a/sandbox/linux-sandbox/3/execroot/envoy/external/io_hyperscan/util -I/tmp/envoy-compdb/_bazel_abeyad/b3914d36ee2d69a6c5998cfece95c54a/sandbox/linux-sandbox/3/execroot/envoy/external/io_hyperscan/src -I/tmp/envoy-compdb/_bazel_abeyad/b3914d36ee2d69a6c5998cfece95c54a/sandbox/linux-sandbox/3/execroot/envoy/bazel-out/k8-fastbuild/bin/contrib/hyperscan/matching/input_matchers/source/hyperscan.build_tmpdir -I/tmp/envoy-compdb/_bazel_abeyad/b3914d36ee2d69a6c5998cfece95c54a/sandbox/linux-sandbox/3/execroot/envoy/external/io_hyperscan -isystem /tmp/envoy-compdb/_bazel_abeyad/b3914d36ee2d69a6c5998cfece95c54a/sandbox/linux-sandbox/3/execroot/envoy/external/io_hyperscan/include -isystem /tmp/envoy-compdb/_bazel_abeyad/b3914d36ee2d69a6c5998cfece95c54a/sandbox/linux-sandbox/3/execroot/envoy/external/org_boost -U_FORTIFY_SOURCE -fstack-protector -Wall -Wthread-safety -Wself-assign -Wunused-but-set-parameter -Wno-free-nonheap-object -fcolor-diagnostics -fno-omit-frame-pointer -std=c++0x -no-canonical-prefixes -Wno-builtin-macro-redefined -D__DATE__=\"redacted\" -D__TIMESTAMP__=\"redacted\" -D__TIME__=\"redacted\" -DABSL_MIN_LOG_LEVEL=4 -fPIC -Wno-deprecated-declarations -std=c++17 -Wno-error=unused-but-set-variable -Wno-error=bitwise-instead-of-logical -march=native -mtune=native -O2 -std=c++11 -Wall -Wextra -Wshadow -Wswitch -Wreturn-type -Wcast-qual -Wno-deprecated -Wnon-virtual-dtor -fno-strict-aliasing -Werror -fno-omit-frame-pointer -MD -MT util/CMakeFiles/expressionutil.dir/expressions.cpp.o -MF util/CMakeFiles/expressionutil.dir/expressions.cpp.o.d -o util/CMakeFiles/expressionutil.dir/expressions.cpp.o -c /tmp/envoy-compdb/_bazel_abeyad/b3914d36ee2d69a6c5998cfece95c54a/sandbox/linux-sandbox/3/execroot/envoy/external/io_hyperscan/util/expressions.cpp
error: unknown warning option '-Werror=bitwise-instead-of-logical' [-Werror,-Wunknown-warning-option]
```

This will be gone if you use LLVM 14.

File: tools/vscode/refresh_compdb.sh [hunk 2]
```diff
@@ -2,11 +2,16 @@
 
 [[ -z "${SKIP_PROTO_FORMAT}" ]] && tools/proto_format/proto_format.sh fix
 
+source_dirs=("//source/..." "//test/..." "//tools/...")
+if [ "$1" == "--include-contrib" ]; then
```
Comment (lizan) [inline tools/vscode/refresh_compdb.sh:6]: lets add this in `gen_compilation_database.py` instead of here.

Comment (abeyad) [reply tools/vscode/refresh_compdb.sh]: Moved to gen_compilation_database.py.  PTAL.

File: tools/gen_compilation_database.py [hunk 3]
```diff
@@ -111,6 +115,7 @@ def fix_compilation_database(args, db):
     parser.add_argument('--include_headers', action='store_true')
     parser.add_argument('--vscode', action='store_true')
     parser.add_argument('--include_all', action='store_true')
+    parser.add_argument('--include_contrib', action='store_true')
```
Comment (lizan) [inline tools/gen_compilation_database.py:118]: Let's default this to true, otherwise it will break clang-tidy in contrib PRs.

Comment (phlax) [reply tools/gen_compilation_database.py:118]: the easiest way to do that it is to flip the switch - eg `--exclude_contrib`

**Review by lizan — APPROVED** [review]

File: tools/gen_compilation_database.py (hunk 3 above)
Comment (abeyad) [reply tools/gen_compilation_database.py]: Changed to `--exclude_contrib`

**Comment by abeyad** [comment]
> Try adding `--copt=-Wno-error,-Wunknown-warning-option` to your bazel command line or `user.bazelrc`?

This worked, thanks for the suggestion!

**Review by lizan — APPROVED** [review]

**Pull request merged**